File: QISHI_Algorithms2/Week2_BFS/problemA1_dfs.java
class Solution {
    public int numIslands(char[][] grid) {
        int n = grid.length;
        if(n == 0)
            return 0;
        int m = grid[0].length;
        int count = 0;
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(grid[i][j] == '1'){
                    dfs(grid, i, j, n, m);
                    count++;
                }
            }
        }
        return count;
    }
    
    public void dfs(char[][] grid, int i, int j, int n, int m){
        grid[i][j] = '0';
        if(i + 1 < n && grid[i + 1][j] == '1')
            dfs(grid, i + 1, j, n, m);
        if(i - 1 >= 0 && grid[i - 1][j] == '1')
            dfs(grid, i - 1, j, n, m);
        if(j + 1 < m && grid[i][j + 1] == '1')
            dfs(grid, i, j + 1, n, m);
        if(j - 1 >= 0 && grid[i][j - 1] == '1')
            dfs(grid, i, j - 1, n, m);
    }
}

File: QISHI_Algorithms2/Week3_DFS/problemA1_DFS.cpp
class Solution {
public:
    vector<unordered_map<int, int>> memo;
    int findTargetSumWays(vector<int>& nums, int S) {
        memo = vector<unordered_map<int, int>>(nums.size());
        return dfs(nums, nums.size(), 0, S);
    }
    
    int dfs(vector<int>& nums, int n, int sp, int S){
        
        if(sp == n){
            if(S == 0) return 1;
            else return 0;
        }
        
        if(memo[sp].count(S))
            return memo[sp][S];
        
        memo[sp][S] = dfs(nums, n, sp + 1, S + nums[sp]) + dfs(nums, n, sp + 1, S - + nums[sp]);
        return memo[sp][S];
    }    
};
